File: typescript/systemdesign/lruCache.ts
/*
In-memory LRU cache with O(1) get/put using a Map (insertion order = recency
order, so touching a key means delete+re-set it to move it to the end).
Common system design interview exercise.

status - completed
*/

import { logp } from "../utils/logger";

class LRUCache<K, V> {
  private readonly capacity: number;
  private readonly cache: Map<K, V> = new Map();

  constructor(capacity: number) {
    if (capacity <= 0) {
      throw new Error("capacity must be positive");
    }
    this.capacity = capacity;
  }

  get(key: K): V | undefined {
    if (!this.cache.has(key)) {
      return undefined;
    }
    const value = this.cache.get(key)!;
    // re-insert to mark as most recently used
    this.cache.delete(key);
    this.cache.set(key, value);
    return value;
  }

  put(key: K, value: V): void {
    if (this.cache.has(key)) {
      this.cache.delete(key);
    } else if (this.cache.size >= this.capacity) {
      // Map preserves insertion order, so the first key is the least recently used
      const lruKey = this.cache.keys().next().value as K;
      this.cache.delete(lruKey);
    }
    this.cache.set(key, value);
  }

  keysInOrder(): K[] {
    return Array.from(this.cache.keys());
  }
}

// usage scenarios

const cache = new LRUCache<number, string>(2);
cache.put(1, "a");
cache.put(2, "b");
logp(`order after inserting 1,2: ${cache.keysInOrder()}`);

logp(`get(1) -> ${cache.get(1)}`);
logp(`order after touching 1: ${cache.keysInOrder()}`);

cache.put(3, "c");
logp("put(3) triggers eviction of least recently used key 2");
logp(`order after inserting 3: ${cache.keysInOrder()}`);
logp(`get(2) -> ${cache.get(2)} (expected undefined, evicted)`);
logp(`get(3) -> ${cache.get(3)}`);

File: python/systemdesign/lru_cache.py
"""
LRU cache with O(1) get/put using a dict of nodes plus a hand-rolled
doubly linked list to track recency order. Classic system design /
interview exercise.

status - completed
"""

from base_logger.logging_event import create_logger


class Node:

    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LruCache:

    log = create_logger(__name__)

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = Node(None, None)
        self.tail = Node(None, None)
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_to_front(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key):
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._remove(node)
        self._add_to_front(node)
        return node.value

    def put(self, key, value):
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self._remove(node)
            self._add_to_front(node)
            return

        if len(self.cache) >= self.capacity:
            lru_node = self.tail.prev
            self._remove(lru_node)
            del self.cache[lru_node.key]

        node = Node(key, value)
        self.cache[key] = node
        self._add_to_front(node)

    def keys_in_order(self):
        order = []
        node = self.head.next
        while node is not self.tail:
            order.append(node.key)
            node = node.next
        return order


if __name__ == "__main__":
    log = create_logger("lru_cache_demo")

    cache = LruCache(capacity=2)
    cache.put(1, "a")
    cache.put(2, "b")
    log.info("order after inserting 1,2: %s", cache.keys_in_order())

    log.info("get(1) -> %s", cache.get(1))
    log.info("order after touching 1: %s", cache.keys_in_order())

    cache.put(3, "c")
    log.info("put(3) triggers eviction of least recently used key 2")
    log.info("order after inserting 3: %s", cache.keys_in_order())
    log.info("get(2) -> %s (expected -1, evicted)", cache.get(2))
    log.info("get(3) -> %s", cache.get(3))
